File: mkkmeans_train.m
% Mehmet Gonen (mehmet.gonen@gmail.com)

function state = mkkmeans_train(Km, parameters)
    tic;
    P = size(Km, 3);
    theta = ones(P, 1) / P;
    K_theta = calculate_kernel_theta(Km, theta.^2);

    opt.disp = 0;
    objective = zeros(parameters.iteration_count, 1);
    for iter = 1:parameters.iteration_count
        fprintf(1, 'running iteration %d...\n', iter);
        [H, ~] = eigs(K_theta, parameters.cluster_count, 'la', opt);

        Q = zeros(P, P);
        for m = 1:P
            Q(m, m) = trace(Km(:, :, m)) - trace(H' * Km(:, :, m) * H);
        end
        res = mskqpopt(Q, zeros(P, 1), ones(1, P), 1, 1, zeros(P, 1), ones(P, 1), [], 'minimize echo(0)');
        theta = res.sol.itr.xx;
        K_theta = calculate_kernel_theta(Km, theta.^2);

        objective(iter) = trace(H' * K_theta * H) - trace(K_theta);
    end
    H_normalized = H ./ repmat(sqrt(sum(H.^2, 2)), 1, parameters.cluster_count);

    stream = RandStream.getGlobalStream;
    reset(stream);
    state.clustering = kmeans(H_normalized, parameters.cluster_count, 'MaxIter', 1000, 'Replicates', 10);
    state.objective = objective;
    state.parameters = parameters;
    state.theta = theta;
    state.time = toc;
end

function K_theta = calculate_kernel_theta(K, theta)
    K_theta = zeros(size(K(:, :, 1)));
    for m = 1:size(K, 3)
        K_theta = K_theta + theta(m) * K(:, :, m);
    end
end

File: mkkmeans_train.R
# Mehmet Gonen (mehmet.gonen@gmail.com)

library(Rmosek)

mkkmeans_train <- function(Km, parameters) {
  state <- list()
  state$time <- system.time({
    P <- dim(Km)[3]
    theta <- rep(1 / P, P)
    K_theta <- matrix(0, nrow(Km), ncol(Km))
    for (m in 1:P) {
      K_theta <- K_theta + theta[m]^2 * Km[,,m]  
    }

    objective <- rep(0, parameters$iteration_count)
    for (iter in 1:parameters$iteration_count) {
      print(sprintf("running iteration %d...", iter))
      H <- eigen(K_theta, symmetric = TRUE)$vectors[, 1:parameters$cluster_count]

      problem <- list()
      problem$sense <- "min"
      problem$c <- rep(0, P)
      problem$A <- Matrix(1, nrow = 1, ncol = P, sparse = TRUE)
      problem$bc <- rbind(blc = 1, buc = 1) 
      problem$bx <- rbind(blx = rep(0, P), bux = rep(1, P))
      problem$qobj <- list(i = 1:P, j = 1:P, v = sapply(1:P, function(m) {sum(diag(Km[,,m])) - sum(diag(t(H) %*% Km[,,m] %*% H))}))
      opts <- list()
      opts$verbose <- 0
      result <- mosek(problem, opts)
      theta <- result$sol$itr$xx
      K_theta <- matrix(0, nrow(Km), ncol(Km))
      for (m in 1:P) {
        K_theta <- K_theta + theta[m]^2 * Km[,,m]  
      }

      objective[iter] <- sum(diag(t(H) %*% K_theta %*% H)) - sum(diag(K_theta))
    }
    H_normalized <- H / matrix(sqrt(rowSums(H^2, 2)), nrow(H), parameters$cluster_count, byrow = FALSE)

    set.seed(NULL)
    state$clustering <- kmeans(H_normalized, centers = parameters$cluster_count, iter.max = 1000, nstart = 10)$cluster
    state$objective <- objective
    state$parameters <- parameters
    state$theta <- theta
  })
  return(state)
}
